### DISCUSSION TITLE
Fetch or import in client component

### DISCUSSION THREAD

**Discussion by Maciekek** [vercel/next.js#61759] [Help] [answered]
### Summary

Hello, 
Today I faced out some problem with choosing approach in my application.

I have client component (List of items with Input. Like in todo list, user can add new item via input, and this item should be automatically visible on the top of the list).

On mount this component i need to load all existing items, so I can do:
- fetch (using for example SWR) 
- import `getAllItems` from server `/api/list.ts` where this function will call DB to get all existing items. 

Which method is better? What is the differences of these methods? I am using nextJs14 and I can't find any docs about this problem. Please help:)

**Comment by icyJoseph** [comment]
Hi, 

Please refer to: https://nextjs.org/docs/app/building-your-application/data-fetching/patterns

I'd say, start with a server component that gets the data directly from the DB, no need for middle ground functions, unless you already have a parametrized function that you can just drop in to fetch the data (as it looks like it is with point 2).

Then pass that data, as props to the client component. Try to filter out and do as many transformations as you can while you are on the server, so that potentially less data is sent to the client, but also less work is required from the client side.

**Reply by Maciekek** [reply]
I know that I will be changing this list a lot with animations, so I want to keep this component as client component. 

I am just thinking is it better to load data via fetch (swr, etc) or via import server function?

**Reply by icyJoseph** [reply]
>  changing this list a lot

To what extend, adding and removing new items from fetch requests? Reorganizing the items? In any rate, you can still create a Server Component that gets the data and passes it as props to your client component, and let it take the wheel from there.

**Reply by Maciekek** [reply]
For example I would like to select specific element from list, after user click on item, i would like to select checkbox visible next to the item, and after 2seconds move this clicked item to the bottom of the list. 
And I am ok to get data from server component and pass them to client component, and there do the magic, but what about adding new items to the list? 

Pass some addItem from server to client component in props as well as items[]? Or send them to API via fetch? 

I am just curious what is best way to do such things, and curious about some limitations of both ways

**Reply by icyJoseph** [reply]
Right, but you can always get "seed" data from the server, and use that as `useState`'s initial value for example, the same would be applicable to a data manager such as SWR or React-Query, and then let the client work on that state. 

Should you sync data with the server, you could use a server action, and even refresh the data on the page, though that will require synching the state via use effect. 

> When you add a new item you can also fire a synchronization action toward the server.

Look at this toy demo, the list is generated server side, a way to append items to that list, in the client, and a counter. We can refresh the data from the server, and the counter will preserve it's state.

Now if I had a DB setup, I could sync through a server action, then upon refresh I would get a correct list.

https://github.com/vercel/next.js/assets/21013447/23a17522-d90b-4c8d-b1fd-3437c3b023d4

Take a look at this Deep Dive document: https://github.com/vercel/next.js/discussions/54075

**Reply by Maciekek** [reply]
Thank you for this conversation, I will use more server components in my solution as you suggest.

 I just wanted to know differences between two approaches, so I am marking @DevDaniloFerrari answer as solution, but your point of view is also important and valuable for me

**Reply by GabenGar** [reply]
You can safely unmark his answer because it's a ChatGPT response, complete with harmful advices, like not using `<form>` element for submitting the data (because a lot of examples of react code feature this anti-pattern), JSX-returning functions which are not React components and not answering the question at all, just rewording your question. Notice how it doesn't mention anything NextJS-specific, like `useSWR()`, because the question which was used in prompt doesn't either (by the virtue of context being obvious from the page).

In regards to [your code sample](https://github.com/vercel/next.js/discussions/61759#discussioncomment-8394786), it won't work at all because `useEffect()` callbacks only run in browser context, so it won't be able to call `auth()` nor `getListByIdQuery()`. 
For the question, you will have to abstract `fetch()` one way or another, specifically URL construction part. Unlike what the answer claims, it's not simple to use out of the box. And also be vary of where you call it, because `fetch()` the browser interface and `fetch()` the nextjs server interface behave differently enough for it to be a problem to debug. So hiding fetch calls behind a specific folder is a good start, if only because it will make relatively simpler to rewrite into the likes of `redux`/`swr`.

**Reply by Maciekek** [reply]
Thanks for the answer, I read this marked answer again, and it really looks like generated by AI :sweat: 

According to your answer, mentioned code actually work (Before write question, I checked it), and maybe because I didn't understand why, I came here with my question.

**Reply by Maciekek** [reply]
Here is a video and code 

https://github.com/vercel/next.js/assets/1099625/35735563-154a-4ba3-b249-500ff90162d4

And code of these two files: 
https://gist.github.com/Maciekek/ece50c2435b2c60412ca686d369f02f5

**Comment by DevDaniloFerrari** [comment]
You can do something like this:

```
"use client";
import { useState } from "react";

export default function Form() {
  const [name, setName] = useState("");
  const [age, setAge] = useState(0);
  const [users, setUsers] = useState([]);

  async function saveUser() {
    await fetch("/api/form", {
      method: "POST",
      body: JSON.stringify({
        name,
        age,
      }),
    });

    const response = await fetch('/api/form')
    const {users} = await response.json()
    setUsers(users)
  }

  function renderUsers() {
    return users?.map((user, index )=> {
        return <li key={index}>{user.name} is {user.age} years old</li>
    })
  }

  return (
    <div>
      <h1>Integrating with API #02</h1>
      <input
        type="text"
        value={name}
        onChange={(event) => setName(event.target.value)}
      />
      <input
        type="number"
        value={age}
        onChange={(event) => setAge(+event.target.value)}
      />
      <button onClick={saveUser}>Submit</button>

      <ul>
        {renderUsers()}
      </ul>
    </div>
  );
}
```

**Reply by Maciekek** [reply]
Yes yes, I know how to do this with fetch and with import method from /api/*.ts, I am just wondering which way is better,
because instead of using `fetch` i can just import
`import { getList, updateList } from 'api/list';` and in `getList`, `updateList` call DB and do operations on DB 


So for example 
File `api/list.ts`

```
'use server'

// imports

export async function getList(listId: number) {
  const session = await auth();

  const result = await getListByIdQuery(session?.user?.email!, listId);
  return result.rows[0]

}
```

And usage of `getList` method in `page.tsx` 

```
'use client';

// imports

export default function List({ params }: { params: { id: number } }) {
  const [status, setStatus] = useState<'loading' | 'ready'>('loading');
  const [list, setList] = useState<{name: string, items: string[]}>();

  useEffect(() => {
    getList(params.id).then((result) => {
      setList({name: result.name, items: result.items} )
      setStatus('ready')
    });
  }, []);

  const addListItem = (e: React.FormEvent<HTMLFormElement>) => {
    e.preventDefault();

    const newList = {
      name: list!.name,
      items: [...list!.items, document.forms[0].elements.itemName.value]
    }

    setList(newList)
    updateList(params.id, newList);

    document.forms[0].elements.itemName.value = ''
  }

  return (
    <div className="flex flex-col h-screen">
      <main className="flex flex-1 flex-col gap-4 p-4 md:gap-8 md:p-6">
        <div className="flex items-center gap-4">
          <form onSubmit={addListItem} className={'w-1/1'}>
            <Input name={'itemName'} className="flex-1" placeholder="Add new item" />
            <Input name={'listId'} className="flex-1 hidden" readOnly value={params.id} />
            <Button variant="outline">Add</Button>
          </form>
        </div>
        <div className="border shadow-sm rounded-lg">
          { status === 'loading' && <div>loading</div>}
          <Table>
            <TableBody>
              { status === 'ready' && list!.items.map((item) => {
                return (
                  <TableRow key={item}>
                    <TableCell className={'w-2'}>
                      <Checkbox id="select-1" />
                    </TableCell>
                    <TableCell className="font-medium">{item}</TableCell>
                  </TableRow>
                );
              })}
            </TableBody>
          </Table>
        </div>
      </main>
    </div>
  );
}

```

**Reply by DevDaniloFerrari** [reply, marked as answer]
### Using `fetch` directly in the component:

**Pros:**
1. **Simplicity:** This approach is straightforward and easy to understand, especially for simple data fetching scenarios.
2. **Component-Level Control:** You have direct control over the fetch logic within the component, making it easier to manage the state locally.

**Cons:**
1. **Code Duplication:** If you need similar fetching logic in multiple components, you might end up duplicating code.
2. **Limited Reusability:** The fetching logic is tied to the specific component, making it less reusable across different parts of your application.

### Importing a function from `/api/*.ts`:

**Pros:**
1. **Code Reusability:** You can centralize your data fetching logic into separate API functions, promoting code reusability across different components.
2. **Abstraction:** Abstracting the fetch logic into separate files can improve code organization and maintainability.
3. **Testing:** Easier unit testing as you can isolate and test the API functions independently.

**Cons:**
1. **Additional Layer:** This approach introduces an additional layer of abstraction, which might be unnecessary for simpler applications.
2. **Complexity for Simple Cases:** For small applications or straightforward data fetching requirements, this might be perceived as over-engineering.

### Recommendation:

For smaller applications or cases where the data fetching logic is straightforward and doesn't require extensive reuse, using `fetch` directly in the component can be a pragmatic choice.

If your application grows, and you find yourself duplicating fetching logic or needing a more organized structure, consider transitioning to the API function approach. It allows for better separation of concerns, promotes code reusability, and simplifies testing.

Ultimately, the "better" approach depends on the specific needs and complexity of your application, as well as your preferences for code organization and maintainability. Consider the trade-offs and choose the method that aligns with your project's requirements and development style.
